src/nupic/algorithms/Connections.cpp

```diff
@@ -82,233 +100,262 @@ Synapse Connections::createSynapse(const Segment& segment,
                                             vector<Cell> input,
                                             SynapseIdx synapseThreshold,
                                             Segment& retSegment) const
 {
-  SynapseIdx numSynapses, maxSynapses = synapseThreshold;
-  vector<SegmentData> segments;
-  vector<SynapseData> synapses;
-  SegmentIdx segmentIdx = 0;
+  SynapseIdx maxActiveSynapses = synapseThreshold;
   bool found = false;
 
   sort(input.begin(), input.end());  // for binary search
 
-  for (auto cell : cells)
+  for (const Cell& cell : cells)
   {
-    segments = cells_[cell.idx].segments;
-    segmentIdx = 0;
+    const vector<SegmentData>& segments = cells_[cell.idx].segments;
 
-    for (auto segment : segments)
+    for (SegmentIdx segmentIdx = 0; segmentIdx < segments.size(); segmentIdx++)
     {
-      synapses = segment.synapses;
-      numSynapses = 0;
-
-      for (auto synapse : synapses)
+      SynapseIdx numActiveSynapses = 0;
+      for (const SynapseData& synapseData : segments[segmentIdx].synapses)
       {
-        if (!synapse.destroyed && synapse.permanence > 0 &&
-            binary_search(input.begin(), input.end(), synapse.presynapticCell))
+        if (!synapseData.destroyed && synapseData.permanence > 0 &&
+            std::binary_search(input.begin(), input.end(),
+                               synapseData.presynapticCell))
         {
-          numSynapses++;
+          numActiveSynapses++;
         }
       }
 
-      if (numSynapses >= maxSynapses)
+      if (numActiveSynapses >= maxActiveSynapses)
       {
-        maxSynapses = numSynapses;
-        retSegment.idx = segmentIdx;
+        maxActiveSynapses = numActiveSynapses;
         retSegment.cell = cell;
+        retSegment.idx = segmentIdx;
         found = true;
       }
-
-      segmentIdx++;
     }
   }
 
   return found;
 }
 
-bool Connections::leastRecentlyUsedSegment(const Cell& cell,
-                                           Segment& retSegment) const
+Segment Connections::leastRecentlyUsedSegment(const Cell& cell) const
 {
+  const vector<SegmentData>& segments = cells_[cell.idx].segments;
   bool found = false;
-  Iteration minIteration = ULLONG_MAX;
-  SegmentData segmentData;
-
-  for (auto segment : segmentsForCell(cell))
+  SegmentIdx minIdx;
+  Iteration minIteration;
+  for (SegmentIdx i = 0; i < segments.size(); i++)
   {
-    // TODO: Possible optimization - define constant variable here?
-    segmentData = dataForSegment(segment);
-
-    if (segmentData.lastUsedIteration < minIteration && !segmentData.destroyed)
+    if (!segments[i].destroyed && (!found ||
+                                   segments[i].lastUsedIteration < minIteration))
     {
-      retSegment = segment;
+      minIdx = i;
+      minIteration = segments[i].lastUsedIteration;
       found = true;
-      minIteration = segmentData.lastUsedIteration;
     }
   }
 
-  return found;
+  NTA_CHECK(found);
+
+  return Segment(minIdx, cell);
```

Comment:
Oh, this method was never giving the caller a reference to anything. It was giving the caller a Segment value (which is just a small struct, not a SegmentData). It did it through an out param rather than as a return value. The params could have also been `(const Cell& cell, Segment* pSegment)` and it could have done `*pSegment = segment`, and the behavior would be equivalent.